compiler/plugins/input/StableHLO/stablehlo-iree/Conversion/StableHLOCustomCalls.cpp

```diff
@@ -0,0 +1,244 @@
+    }
+
+    ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+    auto matrix = op.getOperand(0);
+    auto tau = op.getOperand(1);
+    auto matrixTy = cast<ShapedType>(matrix.getType());
+    auto tauTy = cast<ShapedType>(tau.getType());
+    auto rank = matrixTy.getRank();
+
+    if (isa<ComplexType>(matrixTy.getElementType())) {
+      return rewriter.notifyMatchFailure(op, "complex types not supported");
+    }
+
+    if (rank < 2) {
+      return rewriter.notifyMatchFailure(op, "requires minimum rank 2 matrix");
+    }
+
+    if (!matrixTy.hasStaticShape() || !tauTy.hasStaticShape()) {
+      return rewriter.notifyMatchFailure(op,
+                                         "not supported for dynamic shapes");
+    }
+
+    Value zero = b.create<arith::ConstantIndexOp>(0);
+    Value one = b.create<arith::ConstantIndexOp>(1);
+    Value k = b.create<arith::ConstantIndexOp>(tauTy.getShape().back());
+    Value householder0 = computeHouseholderSlice(matrix, tau, zero, b);
+    auto scf = b.create<scf::ForOp>(
+        one, k, one, ValueRange{householder0},
+        [&](OpBuilder &bb, Location loc, Value iv, ValueRange args) {
+          ImplicitLocOpBuilder b(loc, bb);
+          Value householder = computeHouseholderSlice(matrix, tau, iv, b);
+
+          std::vector<int64_t> batch(rank - 2);
+          for (int i = 0; i < rank - 2; ++i)
+            batch[i] = i;
+          std::vector<int64_t> lhsContract = {rank - 1};
+          std::vector<int64_t> rhsContract = {rank - 2};
+
+          auto dotNums = mlir::stablehlo::DotDimensionNumbersAttr::get(
+              b.getContext(), batch, batch, lhsContract, rhsContract);
+          Value dot = b.create<mlir::stablehlo::DotGeneralOp>(
+              householder0.getType(), args[0], householder, dotNums, nullptr);
+          b.create<scf::YieldOp>(loc, dot);
+        });
+
+    llvm::SmallVector<OpFoldResult> vOffsets(rank, b.getIndexAttr(0));
+    llvm::SmallVector<OpFoldResult> vStrides(rank, b.getIndexAttr(1));
+    llvm::SmallVector<int64_t> vShape(matrixTy.getShape());
+    llvm::SmallVector<OpFoldResult> vSizes;
+    for (auto v : vShape) {
+      vSizes.push_back(b.getIndexAttr(v));
+    }
+
+    auto sliceTy = RankedTensorType::get(vShape, matrixTy.getElementType());
+    Value v = b.create<tensor::ExtractSliceOp>(sliceTy, scf.getResult(0),
+                                               vOffsets, vSizes, vStrides);
+
+    rewriter.replaceOp(op, v);
+    return success();
+  }
+};
+
+//===----------------------------------------------------------------------===//
+// Pass Definition.
+//===----------------------------------------------------------------------===//
+
+struct LegalizeStableHLOCustomCalls final
+    : impl::LegalizeStableHLOCustomCallsBase<LegalizeStableHLOCustomCalls> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<mlir::arith::ArithDialect, mlir::linalg::LinalgDialect,
```

Comment:
You are in mlir namespace already here.

Reply:
Removed for cases where I can remove `mlir::`. The `mlir::stablehlo` one requires the `mlir::` due to some ambiguities between `mlir::` namespace.